text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+```
+
+Since `getOwnConfig` accesses the output of your `configure` build-hook, you have a place to run arbitrary build-time code and inject the results into runtime Javascript.
+
+### Javascript Macro: getConfig
+
+`getConfig` can access the `OwnConfig` of your dependencies.
+
+```js
+import { getConfig } from "@ember/macros";
+const testSelectorsConfig = getConfig("ember-test-selector");
+```
+
+It supports property chaining the same as `getOwnConfig`.
+
+This is a low-level power tool. It's mostly useful as a compile target for custom Babel plugins. For example, `ember-test-selectors` has a custom Babel plugin that _sometimes_ strips test properties out of your components. But if a V2 package is using ember-test-selectors, it needs to run the custom transform _before publishing_. At that point, it's too soon to decide whether to strip them.
+
+Instead of actually doing the stripping, the ember-test-selector plugin would compile the user's code into code that uses `macroCondition` and `getConfig('ember-test-selectors')`. In this way, we get the powerful custom behavior, but only the standard core macros are needed at the time when the app itself is building.
+
+Another example is that a package like `fastboot` could expose config that says whether you're in Fastboot vs the browser, so that apps and addons could say `getConfig('fastboot').isFastboot`. Notice that for an addon to do this, it should declare `fastboot` as an optional peerDependency (see **Optional Peer Dependencies** section), because `getConfig` can only see your **Allowed Dependencies**.
+
+### JavaScript Macro: macroCondition
+
+`macroCondition` acts like a function that takes a boolean value and returns that same boolean value. But whenever `macroCondition` appears directly inside the predicate of an `if` statement or as the predicate of a ternary expression, it tells the macro system to do branch elimination based on the predicate. Here is an example that combines three of the macros we've seen so far:
+
+```js
+// This example:
+import { macroCondition, getOwnConfig, importSync } from "@ember/macros";
+
+let implementation;
+
+if (macroCondition(getOwnConfig().useNewVersion)) {
+  implementation = importSync("./new-component");
+} else {
+  implementation = importSync("./old-component");
+}
+
+export default implementation;
+
+// ==============
+// Could compile down to this if OwnConfig contains { useNewVersion: true }
+let implementation;
+implementation = importSync("./new-component");
+export default implementation;
+
+// ===============
+// Or compile down to this if OwnConfig contains { useNewVersion: false }
+let implementation;
+implementation = importSync("./old-component");
+export default implementation;
+```
+
+It is a build error if `macroCondition` cannot statically determine the truth status of its argument. It's a build error if `macroCondition` appears anywhere other than as the predicate of an if statement or ternary expression.
```

Comment:
Why this restriction? It seems useful to have this lint for things like this:

```js
const isCoolFeatureEnabled = macroCondition(...);
const isCoolFeatureSupported = macroCondition(...);
```

Per my question above, I am not sure if the following is allowed:

```js
if (macroCondition(isCoolFeatureEnabled && isCoolFeatureSupported)) {
  // ...
}
```

I think we should in theory be able to support that with good constant folding. But even if it's not currently supported due to some limitation, I can still do:

```js
if (isCoolFeatureEnabled && isCoolFeatureSupported) {
  // ...
}
```

If I can be sure that the constants were assigned a static boolean value, then even if our macro system can't handle it, I can use some other babel plugin to do the constant folding + dead code elimination. The bottomline is, it seems useful to be able to use `macroCondition` outside of if statements or ternary expressions, since it's just a lint to make sure the expressions are completely static.

Reply:
I think these are good ideas that could improve the ergonomics of the macro system but 
 - they are additive to this RFC
 - the ergonomics of the macro system aren't super high-stakes, because we're still talking about a publication format. People can iterate on more ergonomic tooling that *emits* the macros as output. 

My biggest goals for the macros in this RFC are that we 
 1. don't ship something that we regret needing to support in the future
 2. have reasonable completeness for all the common things addons are doing with custom broccoli today.